File: sdk/python/src/vorpal_sdk/artifact/cpython.py
"""CPython interpreter artifact (python-build-standalone, relocatable).

Mirrors ``sdk/typescript/src/artifact/cpython.js``.

Source name is ``cpython`` (NOT ``python``) — the linux_vorpal bootstrap already
owns a ``python`` source compiled from source, and sources key by (name, platform),
so reusing ``python`` would collide (ADR 0001 / TDD §4, H1).

PROVENANCE — no inline digest (ADR 0001 Part A). The canonical pin is the per-triple
``Vorpal.lock`` entry captured via ``--unlock``; until that lands the HTTP source is
intentionally unpinned (the C1 mint gate fails closed with "unpinned - use --unlock").
A placeholder digest is intentionally avoided — the agent resolves an inline digest
against the registry cache without verifying content, so a predictable placeholder is
a pre-seedable cache-poison key.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from vorpal_sdk.api.artifact import artifact_pb2
from vorpal_sdk.artifact import Artifact, ArtifactSource
from vorpal_sdk.step import shell

if TYPE_CHECKING:
    from vorpal_sdk.context import ConfigContext

# Canonical conforming copy of the build-target CPython interpreter pin. Operator rule:
# CPython 3.13, latest patch. Concrete pin: 3.13.14 (python-build-standalone tag
# 20260623, install_only). Byte-equal to the Rust canonical (ADR 0001 Part B).
DEFAULT_PYTHON_VERSION = "3.13.14"


def cpython_target(system: artifact_pb2.ArtifactSystem) -> str:
    """Map a Vorpal ArtifactSystem to the python-build-standalone target triple.

    Mirrors Rust ``cpython::target()`` — name-agnostic error on unknown system.
    """
    if system == artifact_pb2.AARCH64_DARWIN:
        return "aarch64-apple-darwin"
    if system == artifact_pb2.AARCH64_LINUX:
        return "aarch64-unknown-linux-gnu"
    if system == artifact_pb2.X8664_DARWIN:
        return "x86_64-apple-darwin"
    if system == artifact_pb2.X8664_LINUX:
        return "x86_64-unknown-linux-gnu"
    raise ValueError(f"unsupported toolchain target system: {system}")


class Cpython:
    """Builder for the CPython interpreter artifact."""

    def __init__(self) -> None:
        self._version = DEFAULT_PYTHON_VERSION

    def with_version(self, version: str) -> Cpython:
        self._version = version
        return self

    def build(self, context: ConfigContext) -> str:
        name = "cpython"
        system = context.get_system()

        source_target = cpython_target(system)
        source_version = self._version
        source_path = (
            f"https://sdk.vorpal.build/source/"
            f"cpython-{source_version}-{source_target}.tar.gz"
        )

        source = ArtifactSource(name, source_path).build()

        # pbs install_only tarballs unpack to a top-level python/ dir.
        step_script = f"""mkdir -p "$VORPAL_OUTPUT"
cp -prf "./source/{name}/python/." "$VORPAL_OUTPUT/"
"""
        steps = [shell(context, [], [], step_script, [])]
        systems = [
            artifact_pb2.AARCH64_DARWIN,
            artifact_pb2.AARCH64_LINUX,
            artifact_pb2.X8664_DARWIN,
            artifact_pb2.X8664_LINUX,
        ]

        return (
            Artifact(name, steps, systems)
            .with_aliases([f"{name}:{source_version}"])
            .with_sources([source])
            .build(context)
        )

File: sdk/go/pkg/artifact/cpython.go
package artifact

import (
	"fmt"

	api "github.com/ALT-F4-LLC/vorpal/sdk/go/pkg/api/artifact"
	"github.com/ALT-F4-LLC/vorpal/sdk/go/pkg/config"
)

const defaultPythonVersion = "3.13.14"

// CpythonTarget maps a Vorpal ArtifactSystem to the python-build-standalone target triple.
// Exported for reuse by uv and other python-toolchain artifacts.
func CpythonTarget(system api.ArtifactSystem) (string, error) {
	switch system {
	case api.ArtifactSystem_AARCH64_DARWIN:
		return "aarch64-apple-darwin", nil
	case api.ArtifactSystem_AARCH64_LINUX:
		return "aarch64-unknown-linux-gnu", nil
	case api.ArtifactSystem_X8664_DARWIN:
		return "x86_64-apple-darwin", nil
	case api.ArtifactSystem_X8664_LINUX:
		return "x86_64-unknown-linux-gnu", nil
	default:
		return "", fmt.Errorf("unsupported target system: %s", system.String())
	}
}

// Cpython builds the python-build-standalone relocatable interpreter (install_only).
//
// Source name is "cpython", NOT "python" — the existing linux_vorpal bootstrap owns a
// "python" source compiled from source; sources key by (name, platform), so reusing
// "python" would collide (ADR 0001 / TDD §4, H1).
//
// PROVENANCE — no inline WithDigest (ADR 0001 Part A). The canonical pin is the per-triple
// Vorpal.lock entry captured via --unlock; until then the HTTP source is intentionally
// unpinned and the C1 mint gate fails the build closed. A placeholder digest is
// intentionally avoided: agent.rs returns an inline digest on a registry-cache hit without
// verifying content, so a predictable placeholder is a pre-seedable cache-poison key.
func Cpython(context *config.ConfigContext) (*string, error) {
	name := "cpython"

	system := context.GetTarget()

	sourceTarget, err := CpythonTarget(system)
	if err != nil {
		return nil, err
	}

	sourceVersion := defaultPythonVersion
	sourcePath := fmt.Sprintf(
		"https://sdk.vorpal.build/source/cpython-%s-%s.tar.gz",
		sourceVersion, sourceTarget,
	)

	source := NewArtifactSource(name, sourcePath).Build()

	// pbs install_only tarballs unpack to a top-level python/ dir at the tarball root.
	stepScript := fmt.Sprintf(`mkdir -p "$VORPAL_OUTPUT"
cp -prf "./source/%s/python/." "$VORPAL_OUTPUT/"
`, name)

	step, err := Shell(context, []*string{}, []string{}, stepScript, nil)
	if err != nil {
		return nil, err
	}

	systems := []api.ArtifactSystem{
		api.ArtifactSystem_AARCH64_DARWIN,
		api.ArtifactSystem_AARCH64_LINUX,
		api.ArtifactSystem_X8664_DARWIN,
		api.ArtifactSystem_X8664_LINUX,
	}

	return NewArtifact(name, []*api.ArtifactStep{step}, systems).
		WithAliases([]string{fmt.Sprintf("%s:%s", name, sourceVersion)}).
		WithSources([]*api.ArtifactSource{&source}).
		Build(context)
}
